App peliculas

FrancisLuna opened the pull request:
Se agrega la app de películas
https://trello.com/c/48frPDUV/7-hacer-la-app-de-peliculas:
* Creación de la app películas.
* Creacion de servicios GET, DELETE, POST para películas.
* Creacion de migraciones iniciales para las tablas de peliculas y generos.

FrancisLuna committed:
Se agrega app peliculas
Squashed commit of the following

FrancisLuna committed:
Se vuelve atras un cambio tocado en Examples en testeos.

FrancisLuna committed:
Merge branch 'develop' into fran

FrancisLuna commented on apps/example/migrations/0002_genreentity.py:
Upsi, se me escapo esto en el commit :/

alexiscaspell replied:
Yy bórralo .. , el pr se actualiza solito. Podes seguir modificando sobre el branch con el pr abierto

FrancisLuna commented on apps/peliculas/models.py:
```diff
@@ -0,0 +1,83 @@
+        )
+
+    @staticmethod
+    def buscar_genero_en_DB(genero: Genero) -> "GeneroEntity":
```
Aca no se como pasarle que puede devolver un None sin que chille. -> "GeneroEntity" | None
No le gusta porque no puede hace un | entre un NoneType y un str. Si lo vuelvo str no le gusta que sea str|str ...

alexiscaspell replied:
Con un Optional[GeneroEntity] probaste?

FrancisLuna replied:
Ahi va, no  queria importar el Optional. Supuestamente python 3.10+ lo tenia ... pero ni idea porque chillaba

alexiscaspell commented:
> Se agrega la app de peliculas.

Acá si te interesa hacerlo lo más correcto posible, es buena práctica poner un listado de las funcionalidades que desarrollaste, tipo resumen. Y también se suele meter acá el link del tk de trello relacionado.

FrancisLuna committed:
Arreglo de hint en la funcion.

FrancisLuna committed:
Borraro de una migracion que se genero por error.

alexiscaspell commented on apps/peliculas/models.py:
```diff
@@ -0,0 +1,84 @@
+                genero_query.count() <= 1
+            ), f"Se encontro mas de un genero para el id y nombre dado: {genero_query}"
+            return genero_query.get()
+        # TODO: como verga catcheo el: peliculas.models.GeneroEntity.DoesNotExist: GeneroEntity matching query does not exist. No puedo pasarle el argumento self
```
from django.core.exceptions import ObjectDoesNotExist

alexiscaspell replied:
viste si esto funca @FrancisLuna ?, supuestamente segun habia googleado esa excepcion contiene el DoesNotExist

FabrizioDiPrinzio replied:
@FrancisLuna Proba con `@classmethod` en vez de `@staticmethod`. Los `classmethod` tambien son funciones de clase, pero reciben como argumento la propia clase, por lo cual podrias pasarle `cls.DoesNotExist` en teoria.

Por convencion a los`classmethod` se les pasa como primer parametro `cls` en vez de `self`

FrancisLuna replied:
Se implenta con @classmethod. Gracias :D

alexiscaspell commented on apps/peliculas/models.py:
```diff
@@ -0,0 +1,84 @@
+    def __str__(self) -> str:
+        return f"id: {self.id}, nombre: {self.nombre}"
+
+    def extraer_nuevos_generos(self):
```
por ahi te conviene directamente llamarla nuevos_generos

alexiscaspell commented on apps/peliculas/repository.py:
```diff
@@ -0,0 +1,47 @@
+    pelicula_entity.save()
+    # Segundo se guardan los generos
+    for genero in pelicula.generos:
+        genero_entity = GeneroEntity.buscar_genero_en_DB(genero)
```
aca te quedaria mejor un GeneroEntity.get_by_nombre o como se llame el campo por el que buscas

alexiscaspell replied:
por ahi si es el objeto completo puede ser un get_by_object, no se no tengo mucha imaginacion con los nombres

FrancisLuna replied:
Se cambia a get_by_object.

alexiscaspell commented on apps/peliculas/repository.py:
```diff
@@ -0,0 +1,47 @@
+    search = PeliculaEntity.objects.filter(id=id)
+    if not search.exists():
+        msj = f"Example with id {id} not exist"
+        raise AppException(ExampleError.EXAMPLE_OBJECT_NOT_EXIST, msj)
```
Fijate si aca no queda mejor definir tu propia excepcion que herede de AppException, como PeliculaInexistenteException, o algo asi

FrancisLuna replied:
@brianwolf si creo excepciones propias, donde las meto? en error.py?

FrancisLuna replied:
Se deja como esta >D, alexis contesta jaja

alexiscaspell commented on apps/peliculas/views.py:
```diff
@@ -0,0 +1,26 @@
+
+        movie = service.get(id)
+        if not movie:
+            return HttpResponse(status=204)
```
Esto es medio feucho porque deberia ser un error de negocio (409) o un recurso no encontrado (404), a lo sumo en el peor de los casos un 500. Pero un 204 le estas diciendo: "chabon, ta todo tranca aca pero no hay datos que retornar" y eso esta bien cuando haces una busqueda generica, como por nombre, genero,etc,  pero no cuando accedes por id, xq ahi quiere decir que el que te consulto te envio un id invalido (me entende?).

FabrizioDiPrinzio replied:
cuando el objeto no existe eso siempre es un 404.

FrancisLuna replied:
Se cambia a 404

alexiscaspell reviewed (changes requested):
fran primero revisa esas cositas, igual esta bastante bien el codigo, se entiende y se ve limpio. segui asi :)

FabrizioDiPrinzio commented on apps/peliculas/dto.py:
```diff
@@ -0,0 +1,45 @@
+from peliculas.classes import Genero, Pelicula
+
+
+def genero_a_json(genero: Genero) -> dict[str, object]:
+    return {
+        "id": genero.id,
+        "nombre": genero.nombre,
+    }
+
+
+def json_a_genero(genero_dict: dict[str, object]) -> Genero:
+    return Genero(
+        id=genero_dict.get("id", None),
+        nombre=genero_dict["nombre"],
+    )
+
+
+def pelicula_a_json(pelicula: Pelicula) -> dict[str, object]:
+    return {
+        "id": pelicula.id,
+        "nombre": pelicula.nombre,
+        "fecha_estreno": pelicula.fecha_estreno,
+        "generos": [genero_a_json(genero) for genero in pelicula.generos],
+        "puntuacion": pelicula.puntuacion,
+    }
+
+
+def json_a_pelicula(pelicula_dict: dict[str, object]) -> Pelicula:
+    return Pelicula(
+        id=pelicula_dict.get("id", None),
+        nombre=pelicula_dict["nombre"],
+        fecha_estreno=pelicula_dict["fecha_estreno"],
+        generos=[json_a_genero(genero) for genero in pelicula_dict["generos"]],
+        puntuacion=pelicula_dict["puntuacion"],
+    )
```
Esto podria hacerse con los [Serializers de DJango Rest Framework](https://www.django-rest-framework.org/api-guide/serializers/) para estandarizar como transformar los objetos  jsons y viceversa.

FrancisLuna replied:
Gracias!!!!

FabrizioDiPrinzio commented on apps/peliculas/error.py:
```diff
@@ -0,0 +1,5 @@
+
+
+class ExampleError(Enum):
+    EXAMPLE_OBJECT_NOT_EXIST = 'EXAMPLE_OBJECT_NOT_EXIST'
```
¿Por que el `EXAMPLE_` aca? Yo se lo sacaria porque sino suen muy a demo.

FabrizioDiPrinzio commented on apps/peliculas/migrations/0002_cargar_generos_defecto.py:
```diff
@@ -0,0 +1,35 @@
+
+class Migration(migrations.Migration):
+    dependencies = [("peliculas", "0001_migracion_inicial")]
+    operations = [migrations.RunPython(cargar_generos_default)]
```
el metodo `RunPython` puede recibir un segundo argumento, que es otra funcion que se ejecuta cuando se revierte la migracion. Eso te va a facilitar un monton la vida si estas desarrollando y te das cuenta que hay algo mal con la migracion porque podes volver para atras, hacer los cmabios en las tablas que necesites y volver a correr la migracion para adelante.

Por lo general recomiendo tratar de hacer que tus migraciones sean reversibles si podes porque te va a facilitar un monton la vida.

FrancisLuna replied:
Se agradece!

FabrizioDiPrinzio commented on apps/peliculas/models.py:
```diff
@@ -0,0 +1,84 @@
+    def extraer_nuevos_generos(self):
+        pass
+
+    def to_class(self) -> "Pelicula":
```
En python 3.11, tenes el [typinig.Self](https://peps.python.org/pep-0673/) que podes usar para decir  que un metodo devuelve una instancia de la propia clase. Para considerar si quieren hacer el upgrade para tener acceso a esa feature.

FabrizioDiPrinzio commented on apps/peliculas/repository.py:
```diff
@@ -0,0 +1,47 @@
+
+def list() -> list[int]:
+    search = PeliculaEntity.objects.all()
+    return [pelicula.id for pelicula in search] if search.exists() else []
```
El ternario aca es redundante porque la comprehension ya devuelve una lista vacia si la query no devuelve ningun resultado.

FabrizioDiPrinzio commented on apps/peliculas/repository.py:
```diff
@@ -0,0 +1,47 @@
+    return [pelicula.id for pelicula in search] if search.exists() else []
+
+
+def get(id: int) -> Pelicula:
+    search = PeliculaEntity.objects.filter(id=id)
+    return search.get().to_class() if search.exists() else None
```
Aca podrias hacer simplemente un [PeliculaEntity.objects.filter(id=id).first()](https://peps.python.org/pep-0673/) que ya devuelve el primer resultado o `None` si nada matchea.

FabrizioDiPrinzio commented on apps/peliculas/repository.py:
```diff
@@ -0,0 +1,47 @@
+
+
+def save(pelicula: Pelicula) -> int:
+    # Primero se guarda la pelicula
+    # TODO: Si rompe algo de genero,la pelicula se guarda. Creo que no deberia guardarse...
+    pelicula_entity = PeliculaEntity.from_class(pelicula)
+    pelicula_entity.save()
```
Si no queres que se guarde el objeto si algo se rompe, te conviene hacer el `.save()` al final de la funcion justo antes del return. Asi te aseguras que si hubo algun error antes no guardaste nada.

Tambien podes usar [transactions](https://docs.djangoproject.com/en/4.2/topics/db/transactions/#controlling-transactions-explicitly) que se rollbackean automatica si hay alguna excepcion entonces no guardan los cambios en la base de datos. Recomiendo siempre usar transaccciones cuando vas a tocar algo en la base de datos por las dudas.

FabrizioDiPrinzio reviewed (commented):
Te deje unos comentarios. Algunas son sugerencias, otos son comenarios de cosas de django que pienso que te puedan ayudar, y oras son respuestas a tus preguntas.

FrancisLuna committed:
Correccion de Exception muy generica en buscar_genero_en_DB.

FrancisLuna committed:
Implementacion de @classmethod

Se implementa el decorator @classmethod para catchear un error que es heredado de la clase padre mode.Model. El error es DoesNotExist

FrancisLuna committed:
Bugfix error en guardado de genero

Se corrige para que en caso de que se haya pasado un genero inexistente en la BD,
y dicho genero tiene el nombre y un id inventado, que a la hora de guardar en la BD
se le reasigne automaticamente el ID en vez de usar el que ha pasado el usuario (que
no existe)

FrancisLuna committed:
Se cambia nombre de metodo por uno mas descriptivo.

FrancisLuna committed:
Correccion de status code

FrancisLuna committed:
Actualizacion de nombre de errores.

FrancisLuna committed:
Correccion de nombres de tabla inicial de generos.

FrancisLuna committed:
Se quitan los id de genero a la hora de la busqueda.

Se quita los campos de id en los post.

FrancisLuna committed:
Cambio de nombre por uno mas descriptivo.

FrancisLuna committed:
Refactorizacion de codigo

FrancisLuna committed:
Refactor de codigo.

FrancisLuna committed:
Implementacion de trasacition.

FrancisLuna committed:
uso de walrus

FrancisLuna committed:
Se quita metodo innecesario.

FrancisLuna committed:
Se agrega ignore logs.

FrancisLuna committed:
Merge branch 'develop' into fran

FrancisLuna committed:
Borrado de archivo para posterior hacer test.

FrancisLuna committed:
correcciones de sonar.
